micromasters/serializers_test.py

```diff
@@ -0,0 +1,82 @@
+"""
+Tests for serializing Django User objects
+"""
+import mock
+from django.test import TestCase
+from django.contrib.auth.models import AnonymousUser
+from django.db.models.signals import post_save
+from factory.django import mute_signals
+from micromasters.serializers import UserSerializer
+from micromasters.factories import UserFactory
+from profiles.factories import ProfileFactory
+from search.base import ESTestCase
+# pylint: disable=no-self-use, missing-docstring
+
+
+class UserTests(ESTestCase):
+    """
+    Tests for serializing users.
+    """
+    def test_basic_user(self):
+        with mute_signals(post_save):
+            user = UserFactory.create(email="fake@example.com")
+
+        result = UserSerializer().to_representation(user)
+        assert result == {
+            "username": None,
+            "email": "fake@example.com",
+            "first_name": None,
+            "last_name": None,
+            "preferred_name": None,
+        }
+
+    def test_user_with_profile(self):
+        with mute_signals(post_save):
+            user = UserFactory.create(email="fake@example.com")
+            ProfileFactory.create(
+                user=user,
+                first_name="Rando",
+                last_name="Cardrizzian",
+                preferred_name="Hobo",
+            )
+
+        result = UserSerializer().to_representation(user)
+        assert result == {
+            "username": None,
+            "email": "fake@example.com",
+            "first_name": "Rando",
+            "last_name": "Cardrizzian",
+            "preferred_name": "Hobo",
+        }
+
+    @mock.patch('micromasters.serializers.get_social_username')
+    def test_social_username(self, mock_get_username):
+        mock_get_username.return_value = "remote"
+        with mute_signals(post_save):
+            user = UserFactory.create(
+                username="local", email="fake@example.com",
+            )
+
+        result = UserSerializer().to_representation(user)
+        mock_get_username.assert_called_with(user)
+        assert result == {
+            "username": "remote",
+            "email": "fake@example.com",
+            "first_name": None,
+            "last_name": None,
+            "preferred_name": None,
+        }
+
+
+class AnonymousUserTests(TestCase):
+    """
+    Tests for serializing anonymous users.
+    """
+    def test_serialize_anonymous_user(self):
+        anon_user = AnonymousUser()
+        result = UserSerializer().to_representation(anon_user)
+        assert result is None
+
+    def test_deserialize_anonymous_user(self):
+        result = UserSerializer().to_internal_value(None)
```

Comment:
I spent a fair amount of time researching and trying to figure out the best way to do this, and I'm not sure if I can. Turns out we're not really using DRF's API correctly: the canonical way of serializing an instance to data is:
```python
instance = MyModel()
serializer = MySerializer(instance)
# access `.data` attribute to get the serialized data from the instance
serializer.data
```
And the canonical way of deserializing data is:
```python
data = {"foo": "bar"}
serializer = MySerializer(data=data)
# call `.save()` to turn the data into a model instance and insert/update it in the DB
instance = serializer.save()
```

The `to_representation()` and `to_internal_value()` methods are meant to be overridden when necessary, but never called directly. In addition, when deserializing data, DRF always expects a dictionary -- if I do `UserSerializer(data=None)`, all the built-in validation functions go haywire.

So, I'm not sure what the best course of action is here. I don't think we have any real need to deserialize user objects to Python right now, so we could just take out the `to_internal_value()` method and merge this as-is, perhaps with a comment indicating that this class doesn't support deserializing `None` into an anonymous user. We may want to do a clean-up in the future to use DRF more the way it's meant to be used, but that *definitely* doesn't belong in this pull request.

Thoughts?